File: services/media-service/internal/repository/repository.go
package repository

import (
	"context"
	"encoding/json"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5/pgxpool"

	"github.com/elevatecompact/spark/services/media-service/internal/domain"
)

type MediaRepository struct {
	pool *pgxpool.Pool
}

func NewMediaRepository(pool *pgxpool.Pool) *MediaRepository {
	return &MediaRepository{pool: pool}
}

func (r *MediaRepository) CreateMediaAsset(ctx context.Context, m *domain.MediaAsset) error {
	_, err := r.pool.Exec(ctx, `
		INSERT INTO media_assets (id, uploader_id, content_type, source_filename, file_size_bytes, mime_type, status, storage_path, cdn_url, duration_seconds, width, height, checksum, created_at)
		VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11,$12,$13,$14)
	`, m.ID, m.UploaderID, string(m.ContentType), m.SourceFilename, m.FileSizeBytes, m.MimeType, string(m.Status), m.StoragePath, m.CDNURL, m.DurationSecs, m.Width, m.Height, m.Checksum, m.CreatedAt)
	return err
}

func (r *MediaRepository) GetMediaAsset(ctx context.Context, id uuid.UUID) (*domain.MediaAsset, error) {
	row := r.pool.QueryRow(ctx, `
		SELECT id, uploader_id, content_type, source_filename, file_size_bytes, mime_type, status, storage_path, cdn_url, duration_seconds, width, height, checksum, created_at
		FROM media_assets WHERE id=$1
	`, id)
	m := &domain.MediaAsset{}
	err := row.Scan(&m.ID, &m.UploaderID, &m.ContentType, &m.SourceFilename, &m.FileSizeBytes, &m.MimeType, &m.Status, &m.StoragePath, &m.CDNURL, &m.DurationSecs, &m.Width, &m.Height, &m.Checksum, &m.CreatedAt)
	if err != nil {
		return nil, err
	}
	return m, nil
}

func (r *MediaRepository) UpdateMediaStatus(ctx context.Context, id uuid.UUID, status domain.MediaStatus) error {
	_, err := r.pool.Exec(ctx, `UPDATE media_assets SET status=$2 WHERE id=$1`, id, string(status))
	return err
}

func (r *MediaRepository) UpdateMediaCDN(ctx context.Context, id uuid.UUID, cdnURL string) error {
	_, err := r.pool.Exec(ctx, `UPDATE media_assets SET cdn_url=$2, status='ready' WHERE id=$1`, id, cdnURL)
	return err
}

func (r *MediaRepository) DeleteMediaAsset(ctx context.Context, id uuid.UUID) error {
	_, err := r.pool.Exec(ctx, `UPDATE media_assets SET status='deleted' WHERE id=$1`, id)
	return err
}

func (r *MediaRepository) CreateRendition(ctx context.Context, rd *domain.MediaRendition) error {
	_, err := r.pool.Exec(ctx, `
		INSERT INTO media_renditions (id, media_id, profile, format, file_size_bytes, storage_path, cdn_url, status, created_at)
		VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9)
	`, rd.ID, rd.MediaID, string(rd.Profile), string(rd.Format), rd.FileSizeBytes, rd.StoragePath, rd.CDNURL, string(rd.Status), rd.CreatedAt)
	return err
}

func (r *MediaRepository) GetRenditions(ctx context.Context, mediaID uuid.UUID) ([]domain.MediaRendition, error) {
	rows, err := r.pool.Query(ctx, `
		SELECT id, media_id, profile, format, file_size_bytes, storage_path, cdn_url, status, created_at
		FROM media_renditions WHERE media_id=$1 ORDER BY created_at
	`, mediaID)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var res []domain.MediaRendition
	for rows.Next() {
		var rd domain.MediaRendition
		if err := rows.Scan(&rd.ID, &rd.MediaID, &rd.Profile, &rd.Format, &rd.FileSizeBytes, &rd.StoragePath, &rd.CDNURL, &rd.Status, &rd.CreatedAt); err != nil {
			return nil, err
		}
		res = append(res, rd)
	}
	return res, nil
}

func (r *MediaRepository) CreateTranscodingJob(ctx context.Context, j *domain.TranscodingJob) error {
	profiles, _ := json.Marshal(j.Profiles)
	_, err := r.pool.Exec(ctx, `
		INSERT INTO transcoding_jobs (id, media_id, profiles, status, worker_id, started_at, completed_at, error_message, created_at)
		VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9)
	`, j.ID, j.MediaID, profiles, string(j.Status), j.WorkerID, j.StartedAt, j.CompletedAt, j.ErrorMessage, j.CreatedAt)
	return err
}

func (r *MediaRepository) GetTranscodingJob(ctx context.Context, id uuid.UUID) (*domain.TranscodingJob, error) {
	row := r.pool.QueryRow(ctx, `
		SELECT id, media_id, profiles, status, worker_id, started_at, completed_at, error_message, created_at
		FROM transcoding_jobs WHERE id=$1
	`, id)
	j := &domain.TranscodingJob{}
	var profiles []byte
	err := row.Scan(&j.ID, &j.MediaID, &profiles, &j.Status, &j.WorkerID, &j.StartedAt, &j.CompletedAt, &j.ErrorMessage, &j.CreatedAt)
	if err != nil {
		return nil, err
	}
	json.Unmarshal(profiles, &j.Profiles)
	return j, nil
}

func (r *MediaRepository) UpdateTranscodingJobStatus(ctx context.Context, id uuid.UUID, status domain.TranscodingJobStatus, errMsg string) error {
	_, err := r.pool.Exec(ctx, `UPDATE transcoding_jobs SET status=$2, error_message=$3, completed_at=CASE WHEN $2 IN ('completed','failed') THEN NOW() ELSE NULL END WHERE id=$1`, id, string(status), errMsg)
	return err
}

func (r *MediaRepository) ListTranscodingJobs(ctx context.Context, status *domain.TranscodingJobStatus, limit, offset int) ([]domain.TranscodingJob, error) {
	where := "WHERE 1=1"
	args := []interface{}{}
	idx := 1
	if status != nil {
		where += " AND status=$" + string(rune('0'+idx))
		args = append(args, string(*status))
		idx++
	}
	args = append(args, limit, offset)
	q := `SELECT id, media_id, profiles, status, worker_id, started_at, completed_at, error_message, created_at
	      FROM transcoding_jobs ` + where + ` ORDER BY created_at DESC LIMIT $` + string(rune('0'+idx)) + ` OFFSET $` + string(rune('0'+idx+1))
	rows, err := r.pool.Query(ctx, q, args...)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var res []domain.TranscodingJob
	for rows.Next() {
		var j domain.TranscodingJob
		var profiles []byte
		if err := rows.Scan(&j.ID, &j.MediaID, &profiles, &j.Status, &j.WorkerID, &j.StartedAt, &j.CompletedAt, &j.ErrorMessage, &j.CreatedAt); err != nil {
			return nil, err
		}
		json.Unmarshal(profiles, &j.Profiles)
		res = append(res, j)
	}
	return res, nil
}

func (r *MediaRepository) CreateDRMPolicy(ctx context.Context, p *domain.DRMPolicy) error {
	_, err := r.pool.Exec(ctx, `
		INSERT INTO drm_policies (id, name, content_id, key_system, license_duration_seconds, security_level, is_active, created_at)
		VALUES ($1,$2,$3,$4,$5,$6,$7,$8)
	`, p.ID, p.Name, p.ContentID, string(p.KeySystem), p.LicenseDurationSecs, p.SecurityLevel, p.IsActive, p.CreatedAt)
	return err
}

func (r *MediaRepository) GetDRMPolicies(ctx context.Context) ([]domain.DRMPolicy, error) {
	rows, err := r.pool.Query(ctx, `SELECT id, name, content_id, key_system, license_duration_seconds, security_level, is_active, created_at FROM drm_policies WHERE is_active=true`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var res []domain.DRMPolicy
	for rows.Next() {
		var p domain.DRMPolicy
		if err := rows.Scan(&p.ID, &p.Name, &p.ContentID, &p.KeySystem, &p.LicenseDurationSecs, &p.SecurityLevel, &p.IsActive, &p.CreatedAt); err != nil {
			return nil, err
		}
		res = append(res, p)
	}
	return res, nil
}

func (r *MediaRepository) GetStorageUsage(ctx context.Context) (*domain.StorageUsage, error) {
	var u domain.StorageUsage
	err := r.pool.QueryRow(ctx, `SELECT COALESCE(SUM(file_size_bytes),0) FROM media_assets WHERE status!='deleted'`).Scan(&u.TotalBytes)
	if err != nil {
		return nil, err
	}
	_ = r.pool.QueryRow(ctx, `SELECT COUNT(*) FROM media_assets WHERE status!='deleted'`).Scan(&u.AssetCount)
	return &u, nil
}

func (r *MediaRepository) CheckMediaExists(ctx context.Context, id uuid.UUID) (bool, error) {
	var exists bool
	err := r.pool.QueryRow(ctx, `SELECT EXISTS(SELECT 1 FROM media_assets WHERE id=$1 AND status!='deleted')`, id).Scan(&exists)
	return exists, err
}

func (r *MediaRepository) GetMediaByUploader(ctx context.Context, uploaderID uuid.UUID, limit, offset int) ([]domain.MediaAsset, error) {
	rows, err := r.pool.Query(ctx, `
		SELECT id, uploader_id, content_type, source_filename, file_size_bytes, mime_type, status, storage_path, cdn_url, duration_seconds, width, height, checksum, created_at
		FROM media_assets WHERE uploader_id=$1 AND status!='deleted' ORDER BY created_at DESC LIMIT $2 OFFSET $3
	`, uploaderID, limit, offset)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var res []domain.MediaAsset
	for rows.Next() {
		var m domain.MediaAsset
		if err := rows.Scan(&m.ID, &m.UploaderID, &m.ContentType, &m.SourceFilename, &m.FileSizeBytes, &m.MimeType, &m.Status, &m.StoragePath, &m.CDNURL, &m.DurationSecs, &m.Width, &m.Height, &m.Checksum, &m.CreatedAt); err != nil {
			return nil, err
		}
		res = append(res, m)
	}
	return res, nil
}

func (r *MediaRepository) SaveUploadSession(ctx context.Context, s *domain.UploadSession) error {
	_, err := r.pool.Exec(ctx, `
		INSERT INTO upload_sessions (id, uploader_id, filename, file_size_bytes, content_type, chunks_total, chunks_done, checksum, status, storage_path, created_at, expires_at)
		VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11,$12)
	`, s.ID, s.UploaderID, s.Filename, s.FileSizeBytes, s.ContentType, s.ChunksTotal, s.ChunksDone, s.Checksum, s.Status, s.StoragePath, s.CreatedAt, s.ExpiresAt)
	return err
}

func (r *MediaRepository) GetUploadSession(ctx context.Context, id uuid.UUID) (*domain.UploadSession, error) {
	row := r.pool.QueryRow(ctx, `
		SELECT id, uploader_id, filename, file_size_bytes, content_type, chunks_total, chunks_done, checksum, status, storage_path, created_at, expires_at
		FROM upload_sessions WHERE id=$1
	`, id)
	s := &domain.UploadSession{}
	err := row.Scan(&s.ID, &s.UploaderID, &s.Filename, &s.FileSizeBytes, &s.ContentType, &s.ChunksTotal, &s.ChunksDone, &s.Checksum, &s.Status, &s.StoragePath, &s.CreatedAt, &s.ExpiresAt)
	if err != nil {
		return nil, err
	}
	return s, nil
}

func (r *MediaRepository) UpdateUploadSession(ctx context.Context, id uuid.UUID, chunksDone int, checksum, status string) error {
	_, err := r.pool.Exec(ctx, `UPDATE upload_sessions SET chunks_done=$2, checksum=CASE WHEN $3='' THEN checksum ELSE $3 END, status=$4 WHERE id=$1`, id, chunksDone, checksum, status)
	return err
}

func (r *MediaRepository) DeleteUploadSession(ctx context.Context, id uuid.UUID) error {
	_, err := r.pool.Exec(ctx, `DELETE FROM upload_sessions WHERE id=$1`, id)
	return err
}

File: services/media-service/src/repository.rs
use deadpool_postgres::Pool;
use serde_json;
use tokio_postgres::Row;
use uuid::Uuid;

use crate::domain::*;

pub struct MediaRepository {
    pool: Pool,
}

impl MediaRepository {
    pub fn new(pool: Pool) -> Self {
        Self { pool }
    }

    async fn conn(&self) -> Result<deadpool_postgres::Object, String> {
        self.pool.get().await.map_err(|e| e.to_string())
    }

    pub async fn create_media_asset(&self, m: &MediaAsset) -> Result<(), String> {
        let client = self.conn().await?;
        client
            .execute(
                "INSERT INTO media_assets (id, uploader_id, content_type, source_filename, file_size_bytes, mime_type, status, storage_path, cdn_url, duration_seconds, width, height, checksum, created_at) VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11,$12,$13,$14)",
                &[&m.id, &m.uploader_id, &m.content_type.to_string(), &m.source_filename, &m.file_size_bytes, &m.mime_type, &m.status.to_string(), &m.storage_path, &m.cdn_url, &m.duration_seconds, &m.width, &m.height, &m.checksum, &m.created_at],
            )
            .await
            .map_err(|e| e.to_string())?;
        Ok(())
    }

    pub async fn get_media_asset(&self, id: Uuid) -> Result<MediaAsset, String> {
        let client = self.conn().await?;
        let row = client
            .query_one(
                "SELECT id, uploader_id, content_type, source_filename, file_size_bytes, mime_type, status, storage_path, cdn_url, duration_seconds, width, height, checksum, created_at FROM media_assets WHERE id=$1",
                &[&id],
            )
            .await
            .map_err(|e| e.to_string())?;
        Ok(row_to_media(&row))
    }

    pub async fn update_media_status(&self, id: Uuid, status: &MediaStatus) -> Result<(), String> {
        let client = self.conn().await?;
        client
            .execute("UPDATE media_assets SET status=$2 WHERE id=$1", &[&id, &status.to_string()])
            .await
            .map_err(|e| e.to_string())?;
        Ok(())
    }

    pub async fn update_media_cdn(&self, id: Uuid, cdn_url: &str) -> Result<(), String> {
        let client = self.conn().await?;
        client
            .execute("UPDATE media_assets SET cdn_url=$2, status='ready' WHERE id=$1", &[&id, &cdn_url])
            .await
            .map_err(|e| e.to_string())?;
        Ok(())
    }

    pub async fn delete_media(&self, id: Uuid) -> Result<(), String> {
        let client = self.conn().await?;
        client
            .execute("UPDATE media_assets SET status='deleted' WHERE id=$1", &[&id])
            .await
            .map_err(|e| e.to_string())?;
        Ok(())
    }

    pub async fn create_rendition(&self, r: &MediaRendition) -> Result<(), String> {
        let client = self.conn().await?;
        client
            .execute(
                "INSERT INTO media_renditions (id, media_id, profile, format, file_size_bytes, storage_path, cdn_url, status, created_at) VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9)",
                &[&r.id, &r.media_id, &r.profile, &r.format, &r.file_size_bytes, &r.storage_path, &r.cdn_url, &r.status.to_string(), &r.created_at],
            )
            .await
            .map_err(|e| e.to_string())?;
        Ok(())
    }

    pub async fn get_renditions(&self, media_id: Uuid) -> Result<Vec<MediaRendition>, String> {
        let client = self.conn().await?;
        let rows = client
            .query(
                "SELECT id, media_id, profile, format, file_size_bytes, storage_path, cdn_url, status, created_at FROM media_renditions WHERE media_id=$1 ORDER BY created_at",
                &[&media_id],
            )
            .await
            .map_err(|e| e.to_string())?;
        Ok(rows.iter().map(|row| MediaRendition {
            id: row.get("id"),
            media_id: row.get("media_id"),
            profile: row.get("profile"),
            format: row.get("format"),
            file_size_bytes: row.get("file_size_bytes"),
            storage_path: row.get("storage_path"),
            cdn_url: row.get("cdn_url"),
            status: MediaStatus::from(row.get::<_, String>("status").as_str()),
            created_at: row.get("created_at"),
        }).collect())
    }

    pub async fn create_transcoding_job(&self, j: &TranscodingJob) -> Result<(), String> {
        let client = self.conn().await?;
        let profiles = serde_json::to_value(&j.profiles).unwrap_or_default();
        client
            .execute(
                "INSERT INTO transcoding_jobs (id, media_id, profiles, status, worker_id, started_at, completed_at, error_message, created_at) VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9)",
                &[&j.id, &j.media_id, &profiles, &j.status.to_string(), &j.worker_id, &j.started_at, &j.completed_at, &j.error_message, &j.created_at],
            )
            .await
            .map_err(|e| e.to_string())?;
        Ok(())
    }

    pub async fn get_transcoding_job(&self, id: Uuid) -> Result<TranscodingJob, String> {
        let client = self.conn().await?;
        let row = client
            .query_one("SELECT id, media_id, profiles, status, worker_id, started_at, completed_at, error_message, created_at FROM transcoding_jobs WHERE id=$1", &[&id])
            .await
            .map_err(|e| e.to_string())?;
        Ok(row_to_job(&row))
    }

    pub async fn update_transcoding_job_status(&self, id: Uuid, status: &TranscodingStatus, err_msg: &str) -> Result<(), String> {
        let client = self.conn().await?;
        client
            .execute(
                "UPDATE transcoding_jobs SET status=$2, error_message=$3, completed_at=CASE WHEN $2 IN ('completed','failed') THEN NOW() ELSE NULL END WHERE id=$1",
                &[&id, &status.to_string(), &err_msg],
            )
            .await
            .map_err(|e| e.to_string())?;
        Ok(())
    }

    pub async fn save_upload_session(&self, s: &UploadSession) -> Result<(), String> {
        let client = self.conn().await?;
        client
            .execute(
                "INSERT INTO upload_sessions (id, uploader_id, filename, file_size_bytes, content_type, chunks_total, chunks_done, checksum, status, storage_path, created_at, expires_at) VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11,$12)",
                &[&s.id, &s.uploader_id, &s.filename, &s.file_size_bytes, &s.content_type, &s.chunks_total, &s.chunks_done, &s.checksum, &s.status, &s.storage_path, &s.created_at, &s.expires_at],
            )
            .await
            .map_err(|e| e.to_string())?;
        Ok(())
    }

    pub async fn get_upload_session(&self, id: Uuid) -> Result<UploadSession, String> {
        let client = self.conn().await?;
        let row = client
            .query_one("SELECT id, uploader_id, filename, file_size_bytes, content_type, chunks_total, chunks_done, checksum, status, storage_path, created_at, expires_at FROM upload_sessions WHERE id=$1", &[&id])
            .await
            .map_err(|e| e.to_string())?;
        Ok(UploadSession {
            id: row.get("id"),
            uploader_id: row.get("uploader_id"),
            filename: row.get("filename"),
            file_size_bytes: row.get("file_size_bytes"),
            content_type: row.get("content_type"),
            chunks_total: row.get("chunks_total"),
            chunks_done: row.get("chunks_done"),
            checksum: row.get("checksum"),
            status: row.get("status"),
            storage_path: row.get("storage_path"),
            created_at: row.get("created_at"),
            expires_at: row.get("expires_at"),
        })
    }

    pub async fn update_upload_session(&self, id: Uuid, chunks_done: i32, status: &str) -> Result<(), String> {
        let client = self.conn().await?;
        client
            .execute("UPDATE upload_sessions SET chunks_done=$2, status=$3 WHERE id=$1", &[&id, &chunks_done, &status])
            .await
            .map_err(|e| e.to_string())?;
        Ok(())
    }

    pub async fn delete_upload_session(&self, id: Uuid) -> Result<(), String> {
        let client = self.conn().await?;
        client.execute("DELETE FROM upload_sessions WHERE id=$1", &[&id]).await.map_err(|e| e.to_string())?;
        Ok(())
    }

    pub async fn create_drm_policy(&self, p: &DrmPolicy) -> Result<(), String> {
        let client = self.conn().await?;
        client
            .execute(
                "INSERT INTO drm_policies (id, name, content_id, key_system, license_duration_seconds, security_level, is_active, created_at) VALUES ($1,$2,$3,$4,$5,$6,$7,$8)",
                &[&p.id, &p.name, &p.content_id, &p.key_system, &p.license_duration_seconds, &p.security_level, &p.is_active, &p.created_at],
            )
            .await
            .map_err(|e| e.to_string())?;
        Ok(())
    }

    pub async fn get_drm_policies(&self) -> Result<Vec<DrmPolicy>, String> {
        let client = self.conn().await?;
        let rows = client
            .query("SELECT id, name, content_id, key_system, license_duration_seconds, security_level, is_active, created_at FROM drm_policies WHERE is_active=true", &[])
            .await
            .map_err(|e| e.to_string())?;
        Ok(rows.iter().map(|row| DrmPolicy {
            id: row.get("id"),
            name: row.get("name"),
            content_id: row.get("content_id"),
            key_system: row.get("key_system"),
            license_duration_seconds: row.get("license_duration_seconds"),
            security_level: row.get("security_level"),
            is_active: row.get("is_active"),
            created_at: row.get("created_at"),
        }).collect())
    }

    pub async fn get_storage_usage(&self) -> Result<StorageUsage, String> {
        let client = self.conn().await?;
        let total: i64 = client.query_one("SELECT COALESCE(SUM(file_size_bytes),0) FROM media_assets WHERE status!='deleted'", &[]).await.map_err(|e| e.to_string())?.get(0);
        let count: i64 = client.query_one("SELECT COUNT(*) FROM media_assets WHERE status!='deleted'", &[]).await.map_err(|e| e.to_string())?.get(0);
        Ok(StorageUsage {
            total_bytes: total,
            upload_bytes: total,
            rendition_bytes: 0,
            thumbnail_bytes: 0,
            asset_count: count,
        })
    }

    pub async fn list_transcoding_jobs(&self, status: Option<&str>, limit: i64, offset: i64) -> Result<Vec<TranscodingJob>, String> {
        let client = self.conn().await?;
        let rows = if let Some(s) = status {
            client
                .query("SELECT id, media_id, profiles, status, worker_id, started_at, completed_at, error_message, created_at FROM transcoding_jobs WHERE status=$1 ORDER BY created_at DESC LIMIT $2 OFFSET $3", &[&s, &limit, &offset])
                .await
                .map_err(|e| e.to_string())?
        } else {
            client
                .query("SELECT id, media_id, profiles, status, worker_id, started_at, completed_at, error_message, created_at FROM transcoding_jobs ORDER BY created_at DESC LIMIT $1 OFFSET $2", &[&limit, &offset])
                .await
                .map_err(|e| e.to_string())?
        };
        Ok(rows.iter().map(|row| row_to_job(row)).collect())
    }
}

fn row_to_media(row: &Row) -> MediaAsset {
    MediaAsset {
        id: row.get("id"),
        uploader_id: row.get("uploader_id"),
        content_type: ContentType::from(row.get::<_, String>("content_type").as_str()),
        source_filename: row.get("source_filename"),
        file_size_bytes: row.get("file_size_bytes"),
        mime_type: row.get("mime_type"),
        status: MediaStatus::from(row.get::<_, String>("status").as_str()),
        storage_path: row.get("storage_path"),
        cdn_url: row.get("cdn_url"),
        duration_seconds: row.get("duration_seconds"),
        width: row.get("width"),
        height: row.get("height"),
        checksum: row.get("checksum"),
        created_at: row.get("created_at"),
    }
}

fn row_to_job(row: &Row) -> TranscodingJob {
    let profiles_val: serde_json::Value = row.get("profiles");
    let profiles: Vec<String> = serde_json::from_value(profiles_val).unwrap_or_default();
    TranscodingJob {
        id: row.get("id"),
        media_id: row.get("media_id"),
        profiles,
        status: TranscodingStatus::from(row.get::<_, String>("status").as_str()),
        worker_id: row.get("worker_id"),
        started_at: row.get("started_at"),
        completed_at: row.get("completed_at"),
        error_message: row.get("error_message"),
        created_at: row.get("created_at"),
    }
}
